netbsd/NetBSDProcessList.c

```diff
@@ -0,0 +1,314 @@
+
+   if (opl->kd) {
+      kvm_close(opl->kd);
+   }
+
+   free(opl->cpus);
+
+   ProcessList_done(this);
+   free(this);
+}
+
+static void NetBSDProcessList_scanMemoryInfo(ProcessList* pl) {
+   static int uvmexp_mib[] = {CTL_VM, VM_UVMEXP2};
+   struct uvmexp_sysctl uvmexp;
+   size_t size_uvmexp = sizeof(uvmexp);
+
+   if (sysctl(uvmexp_mib, 2, &uvmexp, &size_uvmexp, NULL, 0) < 0) {
+      CRT_fatalError("uvmexp sysctl call failed");
+   }
+
+   pl->totalMem = uvmexp.npages * pageSizeKB;
+
+   // These calculations have been taken from NetBSD's top(1)
+   // They need review for testing the correctness
+   //pl->freeMem = uvmexp.free * pageSizeKB;
+   pl->buffersMem = uvmexp.filepages * pageSizeKB;
+   pl->cachedMem = (uvmexp.anonpages + uvmexp.filepages + uvmexp.execpages) * pageSizeKB;
+   pl->usedMem = (uvmexp.npages - uvmexp.free - uvmexp.paging) * pageSizeKB + pl->buffersMem + pl->cachedMem;
+
+   pl->totalSwap = uvmexp.swpages * pageSizeKB;
+   pl->usedSwap = uvmexp.swpginuse * pageSizeKB;
+}
+
+static char* NetBSDProcessList_readProcessName(kvm_t* kd, const struct kinfo_proc2* kproc, int* basenameEnd) {
+   /*
+    * Like NetBSD's top(1), we try to fall back to the command name
+    * (argv[0]) if we fail to construct the full command.
+    */
+   char** arg = kvm_getargv2(kd, kproc, 500);
+   if (arg == NULL || *arg == NULL) {
+      *basenameEnd = strlen(kproc->p_comm);
+      return xStrdup(kproc->p_comm);
+   }
+
+   size_t len = 0;
+   for (int i = 0; arg[i] != NULL; i++) {
+      len += strlen(arg[i]) + 1;   /* room for arg and trailing space or NUL */
+   }
+
+   /* don't use xMalloc here - we want to handle huge argv's gracefully */
+   char* s;
+   if ((s = malloc(len)) == NULL) {
+      *basenameEnd = strlen(kproc->p_comm);
+      return xStrdup(kproc->p_comm);
+   }
+
+   *s = '\0';
+
+   for (int i = 0; arg[i] != NULL; i++) {
+      size_t n = strlcat(s, arg[i], len);
+      if (i == 0) {
+         *basenameEnd = MINIMUM(n, len - 1);
+      }
+      /* the trailing space should get truncated anyway */
+      strlcat(s, " ", len);
+   }
+
+   return s;
+}
+
+/*
+ * Borrowed with modifications from NetBSD's top(1).
+ */
+static double getpcpu(const struct kinfo_proc2* kp) {
+   if (fscale == 0)
+      return 0.0;
+
+   return 100.0 * (double)kp->p_pctcpu / fscale;
+}
+
+static void NetBSDProcessList_scanProcs(NetBSDProcessList* this) {
+   const Settings* settings = this->super.settings;
+   bool hideKernelThreads = settings->hideKernelThreads;
+   bool hideUserlandThreads = settings->hideUserlandThreads;
+   int count = 0;
+   int nlwps = 0;
+
+   const struct kinfo_proc2* kprocs = kvm_getproc2(this->kd, KERN_PROC_ALL, 0, sizeof(struct kinfo_proc2), &count);
+
+   for (int i = 0; i < count; i++) {
+      const struct kinfo_proc2* kproc = &kprocs[i];
+
+      bool preExisting = false;
+      Process* proc = ProcessList_getProcess(&this->super, kproc->p_pid, &preExisting, NetBSDProcess_new);
```

Comment:
Several fields in the global `Process` structure will be renamed following #388. This mostly affects the `cmdline`, `procComm` (Custom Thread/command name) and `procExe` (main executable file reported by the kernel). When any of these fields is updated a call to `Process_makeCommandStr` needs to be done to update some internal process data. Also note that the thread handling is now handled in the global `Process` structure via the `isUserlandThread` and `isKernelThread` fields. Implementations should set those fields accordingly (`isUserlandThread` should be `false` for the main process)

Reply:
Should be updated in 237e66c